File: src/main/java/io/springreact/web/ReactView.java
package io.springreact.web;

import java.util.Map;

import io.springreact.autoconfigure.ReactRenderer;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

import org.springframework.web.servlet.View;

/**
 * A Spring MVC {@link View} backed by React. Analogous to a Thymeleaf template view:
 * given the model produced by the controller, it renders the React HTML shell and
 * writes it to the response.
 */
public class ReactView implements View {

    private final String viewName;
    private final ReactRenderer renderer;

    public ReactView(String viewName, ReactRenderer renderer) {
        this.viewName = viewName;
        this.renderer = renderer;
    }

    @Override
    public String getContentType() {
        return "text/html;charset=UTF-8";
    }

    @Override
    public void render(Map<String, ?> model, HttpServletRequest request,
                       HttpServletResponse response) throws Exception {
        response.setContentType(getContentType());
        String html = renderer.render(viewName, model);
        response.getWriter().write(html);
    }
}

File: src/main/kotlin/io/springreact/web/ReactView.kt
package io.springreact.web

import io.springreact.autoconfigure.ReactRenderer
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.core.Ordered
import org.springframework.web.servlet.View
import org.springframework.web.servlet.ViewResolver
import java.util.Locale

/** A Spring MVC [View] backed by a server component — the Thymeleaf-style shell. */
class ReactView(private val viewName: String, private val renderer: ReactRenderer) : View {

    override fun getContentType() = "text/html;charset=UTF-8"

    override fun render(model: Map<String, *>?, request: HttpServletRequest, response: HttpServletResponse) {
        response.contentType = contentType
        response.writer.write(renderer.render(viewName, model ?: emptyMap<String, Any?>()))
    }
}

/** Resolves any controller-returned view name into a [ReactView] (fallback resolver). */
class ReactViewResolver(private val renderer: ReactRenderer) : ViewResolver, Ordered {

    override fun resolveViewName(viewName: String, locale: Locale): View? {
        if (viewName.startsWith("redirect:") || viewName.startsWith("forward:") || viewName == "error") return null
        return ReactView(viewName, renderer)
    }

    override fun getOrder() = Ordered.LOWEST_PRECEDENCE - 10
}
